blog/_posts/2016-12-xx-moredots.md

```diff
@@ -0,0 +1,449 @@
+has performance comparable to C or Fortran, so there is no *necessity*
+of vectorizing, but of course vectorization may still be convenient for some problems.
+And vectorized operations like `scalar*array` or `sqrt(array)` are still fast in Julia
+(calling optimized library routines, albeit ones written in Julia itself).
+
+Furthermore, if your problem involves a function that does not have a pre-written,
+highly optimized, vectorized, library routine in Julia, and does not
+decompose easily into existing vectorized building blocks like `scalar*array`, then
+you can write your own building block without dropping down to a low-level language.
+(If all the performance-critical code you will ever need already existed in the
+form of optimized library routines, programming would be a lot easier!)
+
+### Why vectorized code is not as fast as it could be
+
+There is a tension between two general principles in computing: on
+the one hand, re-using highly optimized code is often good for
+performance; on the other other hand, optimized code that is specialized
+for your problem can usually beat general-purpose functions.
+This is illustrated nicely by the vectorized version of our code above:
+
+```jl
+f(x) = 3x.^2 + 5x + 2
+X = f(2 * X.^2 + 6 * X.^3 - sqrt(X))
+```
+
+Each of the operations like `X.^2`  and `5*X` *individually*
+calls highly optimized functions, but their *combination*
+leaves a lot of performance on the table when `X` is an array.   To see that,
+you have to realize that this code is equivalent to:
+
+```
+tmp1 = X.^2
+tmp2 = 2*tmp1
+tmp3 = X.^3
+tmp4 = 6 * tmp3
+tmp5 = tmp2 + tmp4
+tmp6 = sqrt(X)
+tmp7 = tmp5 - tmp7
+X = f(tmp7)
+```
+
+That is, each of these vectorized operations allocates a separate
+temporary array, and is a separate library call with its own inner
+loop.  Both of these properties are bad for performance.
+
+First, eight arrays are allocated (`tmp1` through `tmp7`, plus another
+for the result of `f(tmp7)`, and another four are allocated
+internally by `f(tmp7)` for the same reasons, for *12 arrays in all*.
+The resulting `X = ...` expression does *not* update `X` in-place, but
+rather makes the variable `X` "point" to a new array returned by `f(tmp7)`,
+discarding the old array `X`.   All of these extra arrays are eventually
+deallocated by Julia's garbage collector, but in the meantime it wastes
+a lot of memory (more than an order of magnitude!)
+
+If the array `X` is small, then the performance cost of allocating
+these temporary arrays is significant: heap allocation is expensive.
+If the array `X` is large enough, then the time for the heap allocation
+and garbage collection may be negligible, but you pay a *different* performance price
+from the fact that you have 12 loops (12 passes over memory) compared
+to one, largely because of the loss of [cache locality](https://en.wikipedia.org/wiki/Locality_of_reference).
+
+In particular, main computer memory (RAM) is relatively slow (much slower than
```

Comment:
I think if I didn't already feel comfortable with this material, this section would confuse me badly because I'd assume that temporal locality would only benefit me if the cost of computing all of my fused operations on `X[i]` were small relative to the cost of loading `X[i]` from main memory into the cache. But it might not have been obvious to me that adding two numbers together is so much less expensive than loading them into cache.